File: n_gram_model.py
import argparse
import random
import sys
import typing
from typing import Dict, List, Tuple


class FreqTable:
    def __init__(self, word: str = None):
        self.table = {word: 1} if word else {}
        self.n_observations = len(self.table)

    def observed(self, word: str):
        if word in self.table:
            self.table[word] += 1
        else:
            self.table[word] = 1
        self.n_observations += 1

    def get_random_word(self) -> str:

        # Only one candidate word in table; simply return that one (simple optimization)
        if self.n_observations == 1:
            return next(iter(self.table.keys()))

        # Multiple candidates; select randomly, weighted by probability
        r = random.randrange(self.n_observations)
        cumul = 0
        for next_word, count in self.table.items():
            cumul += count
            if r < cumul:
                return next_word

        return "."  # Something went wrong, return "." to end sentence.


def read_text_corpus(filename: str) -> List[str]:

    # Read the input data, and turn it into a list of words.
    with open(filename) as f:
        lines = f.readlines()
    return " ".join(lines).split(" ")


def train_n_gram_model(text_corpus: List[str], n: int) -> Dict[str, FreqTable]:

    n_gram_model: Dict[str, FreqTable] = {}

    for ww in range(len(text_corpus) - n):

        # Use the n words as key, and the n+1 word as the value to map to
        n_gram_key_str = " ".join(text_corpus[ww : ww + n])
        next_word = text_corpus[ww + n]

        if not n_gram_key_str in n_gram_model:
            # New n_gram: initialize frequency table for it.
            n_gram_model[n_gram_key_str] = FreqTable(next_word)
        else:
            # n_gram was already observed: update frequency table.
            n_gram_model[n_gram_key_str].observed(next_word)

    return n_gram_model


def generate_sentence(n_gram_model: Dict[str, FreqTable], n: int, seed_words: List[str]) -> str:

    # Start sentence with the seed_words (deep copy)
    words = [w for w in seed_words]

    # Stop once the last word is a punctuation mark
    while not words[-1] in [".", "?", "!"]:

        # Convert last n words to one string, i.e. the key
        n_gram_key_str = " ".join(words[-n:])

        next_word = "."  # Default: end sentence if no next_word found
        if n_gram_key_str in n_gram_model:
            next_word = n_gram_model[n_gram_key_str].get_random_word()

        words.append(next_word)

    return " ".join(words)


def process_arguments() -> Tuple[str, int, List[str]]:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "corpus_filename",
        default="data/sherlock_holmes.txt",
        nargs="?",
        help="file (txt) or directory to read cost vars from",
    )
    parser.add_argument("n", type=int, default=2, nargs="?", help="length of the n-gram, i.e. 'n'")
    parser.add_argument(
        "seed_sentence", default="i confess", nargs="?", help="seed sentence (with #words >= n)"
    )
    args = parser.parse_args()

    seed_words = args.seed_sentence.split(" ")
    if len(seed_words) < args.n:
        raise Exception(
            f"'seed_sentence' should have at least {args.n} words, but '{args.seed_sentence}' has length {len(seed_words)}."
        )

    return args.corpus_filename, args.n, seed_words


def main():

    corpus_filename, n, seed_words = process_arguments()

    text_corpus = read_text_corpus(corpus_filename)

    n_gram_model = train_n_gram_model(text_corpus, n)

    for _ in range(10):
        print(generate_sentence(n_gram_model, n, seed_words))


if __name__ == "__main__":
    main()

File: n_gram_model.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

class n_gram_model {

  static class FreqTable {
    private HashMap<String, Integer> table;
    private int n_observations;
    Random random = new Random();

    public FreqTable(String word) {
      this.table = new HashMap<>();
      this.table.put(word, 1);
      this.n_observations = 1;
    }

    public FreqTable() {
      this.table = new HashMap<>();
      this.n_observations = 0;
    }

    public void observed(String next_word) {
      if (this.table.containsKey(next_word)) {
        int val = this.table.get(next_word);
        this.table.put(next_word, val + 1);
      } else {
        this.table.put(next_word, 1);
      }
      this.n_observations += 1;
    }

    public String get_random_word() {

      // Only one candidate word in table; simply return that one (simple optimization)
      if (this.table.size() == 1) this.table.entrySet().iterator().next().getKey();

      // Multiple candidates; select randomly, weighted by probability
      int r = random.nextInt(this.n_observations);
      int cumul = 0;
      for (String next_word : table.keySet()) {
        cumul += table.get(next_word);
        if (r < cumul) return next_word;
      }

      return new String("."); // Something went wrong, return "." to end sentence.
    }

    public String toString() { // Exception: must be camelCased to blend with println()
      return table.toString();
    }
  }

  static HashMap<String, FreqTable> train_n_gram_model(List<String> text_corpus, int n) {
    HashMap<String, FreqTable> n_gram_model_obj = new HashMap<>();

    for (int ww = 0; ww < text_corpus.size() - n; ww++) {

      String n_gram_key_str = new String();
      for (int i = 0; i < n; i++) n_gram_key_str += text_corpus.get(ww + i) + " ";
      String next_word = text_corpus.get(ww + n);

      if (!n_gram_model_obj.containsKey(n_gram_key_str)) {
        n_gram_model_obj.put(n_gram_key_str, new FreqTable(next_word));
      } else {
        n_gram_model_obj.get(n_gram_key_str).observed(next_word);
      }
    }

    return n_gram_model_obj;
  }

  static String generate_sentence(
      HashMap<String, FreqTable> n_gram_model_obj, int n, List<String> seed_words) {

    // Start sentence with the seed_words (deep copy)
    ArrayList<String> words = new ArrayList<>(seed_words);

    // Stop once the last word is a punctuation mark
    List<String> punctuation_marks = Arrays.asList(".", "!", "?");
    while (!punctuation_marks.contains(words.get(words.size() - 1))) {

      // Convert last n words to one string, i.e. the key
      String n_gram_key_str = new String();
      for (int i = -n; i < 0; i++) n_gram_key_str += words.get(words.size() + i) + " ";

      String next_word = new String("."); // Default: end sentence if no next_word found
      if (n_gram_model_obj.containsKey(n_gram_key_str)) {
        FreqTable freq_table = n_gram_model_obj.get(n_gram_key_str);
        next_word = freq_table.get_random_word();
      }
      words.add(next_word);
    }
    return String.join(" ", words);
  }

  static List<String> read_text_corpus(String filename) {
    try {
      String content = Files.readString(Path.of(filename));
      return Arrays.asList(content.split(" "));
    } catch (IOException ex) {
      throw new RuntimeException("ERROR: Could not find filename '" + filename + "'");
    }
  }

  static class Arguments {
    public final String filename;
    public final int n;
    public final List<String> seed_words;

    public Arguments(String filename, int n, List<String> seed_words) {
      this.filename = filename;
      this.n = n;
      this.seed_words = seed_words;
    }
  }

  static Arguments process_arguments(String[] args) {
    String filename = (args.length > 0 ? args[0] : new String("data/sherlock_holmes.txt"));
    int n = (args.length > 1 ? Integer.parseInt(args[1]) : 2);
    String seed_sentence = (args.length > 2 ? args[2] : "i confess");
    List<String> seed_words = Arrays.asList(seed_sentence.split(" "));

    if (seed_words.size() < n)
      throw new RuntimeException(
          "'seed_sentence' should have at least "
              + n
              + " words, but '"
              + seed_sentence
              + "' has length "
              + seed_words.size()
              + ".");

    return new Arguments(filename, n, seed_words);
  }

  public static void main(String[] args_array) {

    Arguments args = process_arguments(args_array);

    List<String> text_corpus = read_text_corpus(args.filename);

    HashMap<String, FreqTable> n_gram_model_obj = train_n_gram_model(text_corpus, args.n);

    for (int s = 0; s < 10; s++)
      System.out.println(generate_sentence(n_gram_model_obj, args.n, args.seed_words));
  }
}
